File: ValidParenthesesString/valid_parentheses_string.py
"""
RULES:
    - Any left parenthesis '(' must have a corresponding right parenthesis ')'.
    - Any right parenthesis ')' must have a corresponding left parenthesis '('.
    - Left parenthesis '(' must go before the corresponding right parenthesis ')'.
    - '*' could be treated as a single right parenthesis ')' or a single left parenthesis '(' or an empty string.
    - An empty string is also valid.

"""class Solution(object):
    def checkValidString(self, s):
        """
        :type s: str
        :rtype: bool
        """
        left, star = [], []
        for i in range (len(s)):
            if s[i] == '(':
                left.append(i)
            elif s[i] == '*':
                star.append(i)
            else:
                # we have encountered a closing bracket
                # gotta pop now, from left first if it
                # has something, otherwise pop from star
                if len(left) == 0 and len(star) == 0: return False
                if len(left) != 0:
                    left.pop()
                else:
                    star.pop()
        # STORY TIME
        # now for cases we do not encounter a closing bracket at all. such cases
        # won't be taken care by the three conditional statements above but our
        # stacks contain some stuff at this point, using which we gotta decide
        # the validity of the string. 
        #
        # So we pop and compare the indices. Indices
        # stored in left can never be greater than that stored in start. WHY?
        #
        # take a case **((. In this left will have 2,3 and star will have 0,1. 
        # on popping se see that left.pop() > star.pop() and we know that its not
        # valid. as it can result in (( or ))(( or ((((. all of which are not
        # valid.
        while(len(left) != 0 and len(star) != 0):
            if(left.pop() > star.pop()):
                return False
        # finally, a correctly balanced string would result in empty left stack
        return len(left) == 0

File: ValidParenthesesString/valid_parentheses_string.java
/*
RULES:
	- Any left parenthesis '(' must have a corresponding right parenthesis ')'.
	- Any right parenthesis ')' must have a corresponding left parenthesis '('.
	- Left parenthesis '(' must go before the corresponding right parenthesis ')'.
	- '*' could be treated as a single right parenthesis ')' or a single left parenthesis '(' or an empty string.
	- An empty string is also valid.
*/

class Solution {
     public boolean checkValidString(String s) {
        /* keep separate stacks for opening bracket
         * and star. When closing is encountered, pop
         * from leftID first if not empty. if leftID
         * is empty, pop from starID if not empty. if
         * both of them are empty, we have a false*/

        Stack<Integer> leftID = new Stack<>();
        Stack<Integer> starID = new Stack<>();

         for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(')
                leftID.push(i);
            else if (ch == '*')
                starID.push(i);
            else {
                if (leftID.isEmpty() && starID.isEmpty())   return false;
                if (!leftID.isEmpty())
                    leftID.pop();
                else
                    starID.pop();
            }
        }
        /* this is for cases where we don't encounter a closing
         * at all. such as "((*(*(**". Here the key is that index
         * stored in leftID must always be less than index stored in
         * starID. This is because opening cannot appear after last
         * star (there is no way to balance that)*/
        while (!leftID.isEmpty() && !starID.isEmpty()) {
            if (leftID.pop() > starID.pop())
                return false;
        }
        /* correct balanced sequence will result in an empty leftID */
        return leftID.isEmpty();
    }
}
